crates/ra_ssr/src/parsing.rs

```diff
@@ -0,0 +1,238 @@
+//! This file contains code for parsing SSR rules, which look something like `foo($a) ==>> bar($b)`.
+//! We first split everything before and after the separator `==>>`. Next, both the search pattern
+//! and the replacement template get tokenized by the Rust tokenizer. Tokens are then searched for
+//! placeholders, which start with `$`. For replacement templates, this is the final form. For
+//! search patterns, we go further and parse the pattern as each kind of thing that we can match.
+//! e.g. expressions, type references etc.
+
+use crate::{
+    PatternElement, Placeholder, RawSearchPattern, SsrError, SsrPattern, SsrRule, SsrTemplate,
+    Token,
+};
+use ra_syntax::{ast, AstNode, SmolStr, SyntaxKind};
+use rustc_hash::FxHashSet;
+use std::{collections::HashMap, str::FromStr};
```

Comment:
Similarly, use `FxHashMap`. They are faster and deterministic. It's also good to stick to one kind of hashmap, to reduce monomorphisations.